### PR TITLE
(JS) Implement parallel walk io

### PR DESCRIPTION
**Pull request by rafael-atticlabs**
Fixes: #3068

### CODE REVIEW

File: js/noms/package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@attic/noms",
   "license": "Apache-2.0",
-  "version": "67.1.0",
+  "version": "68.0.0",
```
Comment (rafael-atticlabs) [inline js/noms/package.json]: Note I bumped the major version since the semantics changed on ValueBase.walkValues

**Comment by rafael-atticlabs** [comment]
jenkins: test this

File: js/noms/src/walk.js [hunk 2]
```diff
@@ -18,24 +24,78 @@ export type WalkCallback = (v: Value) => ?boolean | Promise<?boolean>;
  * this node's children.
  *
  * If `cb` returns undefined or `Promise.resolve()`, the default is to continue recursing (`false`).
```
Comment (arv) [inline js/noms/src/walk.js]: This comment is not correct any more.

File: js/noms/src/walk.js [hunk 3]
```diff
@@ -18,24 +24,78 @@ export type WalkCallback = (v: Value) => ?boolean | Promise<?boolean>;
...
+ *
  */
+
+interface ValueRec {
```
Comment (arv) [inline js/noms/src/walk.js]: Use type here? Also, what is cb?

```js
type ValueRec = {
  value: Value,
  done: boolean,
};
```

Comment (arv) [inline js/noms/src/walk.js]: cb -> needsCallback?

Comment (rafael-atticlabs) [inline js/noms/src/walk.js]: done

File: js/noms/src/walk.js (hunk 2 above)
Comment (rafael-atticlabs) [inline js/noms/src/walk.js]: good catch. now it is (switched to existing semantics).

File: js/noms/src/value.js [hunk 4]
```diff
@@ -9,9 +9,10 @@ import type {primitive} from './primitives.js';
 import {ensureHash} from './get-hash.js';
 import type {Type} from './type.js';
 import type Ref from './ref.js';
-import type {WalkCallback} from './walk.js';
 import type {ValueReader} from './value-store.js';
 
+export type ValueCallback = (v: Value) => ?void | Promise<?void>;
```
Comment (arv) [inline js/noms/src/value.js]: This type is strange. `?T` means `null | void | T` so `?void` is the same as `null | void | void` Do we want to allow `null` or not?

File: js/noms/src/walk.js [hunk 5]
```diff
@@ -18,24 +24,84 @@ export type WalkCallback = (v: Value) => ?boolean | Promise<?boolean>;
...
-    case 'string':
-      return;
-  }
+  while (values.length > 0 || refs.length > 0) {
+    // Visit all values located *within* loaded chunks.
+    while (values.length > 0) {
+      const rec = values.pop();
+      if (rec.needsCallback) {
+        let skip = cb(rec.v);
+        if (skip && skip !== true) {
```
Comment (arv) [inline js/noms/src/walk.js]: `skip` is either undefined or null so `skip !== true` is always true.

**Pull request merged**